src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/GoogleDataprocDAO.scala

```diff
@@ -199,27 +188,137 @@ class GoogleDataprocDAO(protected val dataprocConfig: DataprocConfig, protected
       StorageObjectResponse(storageObject.getName, storageObject.getBucket, storageObject.getTimeCreated.toString)
     }
   }
 
   /* Delete a cluster within the google project */
   def deleteCluster(googleProject: String, clusterName: String)(implicit executionContext: ExecutionContext): Future[Unit] = {
-    Future {
-      val request = dataproc.projects().regions().clusters().delete(googleProject, dataprocConfig.dataprocDefaultZone, clusterName)
-      try {
-        executeGoogleRequest(request)
-      } catch {
-        case e:GoogleJsonResponseException =>
-          if(e.getStatusCode!=404)
-            throw CallToGoogleApiFailedException(googleProject, clusterName, e.getStatusCode, e.getDetails.getMessage)
+    val request = dataproc.projects().regions().clusters().delete(googleProject, dataprocConfig.dataprocDefaultRegion, clusterName)
+    executeGoogleRequestAsync(googleProject, clusterName, request).recover {
+      // treat a 404 error as a successful deletion
+      case CallToGoogleApiFailedException(_, _, 404, _) => ()
+    }.void
+  }
+
+  override def getClusterStatus(googleProject: GoogleProject, clusterName: String)(implicit executionContext: ExecutionContext): Future[LeoClusterStatus] = {
+    getCluster(googleProject, clusterName).map { cluster =>
+      LeoClusterStatus.withNameIgnoreCase(cluster.getStatus.getState)
+    }
+  }
+
+  override def getClusterMasterInstanceIp(googleProject: GoogleProject, clusterName: String)(implicit executionContext: ExecutionContext): Future[Option[String]] = {
+    // OptionT is handy when you potentially want to deal with Future[A], Option[A],
+    // Future[Option[A]], and A all in the same flatMap!
+    //
+    // Legend:
+    // - OptionT.pure turns an A into an OptionT[F, A]
+    // - OptionT.liftF turns an F[A] into an OptionT[F, A]
+    // - OptionT.fromOption turns an Option[A] into an OptionT[F, A]
+
+    val ipOpt: OptionT[Future, String] = for {
+      cluster <- OptionT.liftF[Future, GoogleCluster] { getCluster(googleProject, clusterName) }
+      masterInstanceName <- OptionT.fromOption { getMasterInstanceName(cluster) }
+      masterInstanceZone <- OptionT.fromOption { getZone(cluster) }
+      masterInstance <- OptionT.liftF[Future, Instance] { getInstance(googleProject, masterInstanceZone, masterInstanceName) }
+      masterInstanceIp <- OptionT.fromOption { getInstanceIP(masterInstance) }
+    } yield masterInstanceIp
+
+    // OptionT[Future, String] is simply a case class wrapper for Future[Option[String]].
+    // So this just grabs the inner value and returns it.
+    ipOpt.value
+  }
+
+  override def getClusterErrorDetails(operationName: String)(implicit executionContext: ExecutionContext): Future[Option[ClusterErrorDetails]] = {
+    val errorOpt: OptionT[Future, ClusterErrorDetails] = for {
+      operation <- OptionT.liftF[Future, DataprocOperation] { getOperation(operationName) }
+      _ <- OptionT.pure { operation.getDone }.filter(_ == true)
```

Comment:
Maybe this is the point where it would be cleaner to use `flatMap`s.  This line is very confusing.

Reply:
Maybe - I've rewritten this line a couple times now. Basically I want to abort the flatMap if operation.getDone == false. I can see how it would look as a flatMap.

Reply:
Or I can just remove the line!
```
  val errorOpt: OptionT[Future, ClusterErrorDetails] = for {
      operation <- OptionT.liftF[Future, DataprocOperation] { getOperation(operationName) } if operation.getDone
      error <- OptionT.pure { operation.getError }
      code <- OptionT.pure { error.getCode }
    } yield ClusterErrorDetails(code, Option(error.getMessage))
```

Reply:
nice!